configurations/configuration.py

```diff
@@ -53,48 +54,172 @@ def set_database_connection_variables(config):
+    ):
+        for data_source in configuration.get("DATA_SOURCES"):
+            if (
+                data_source.get("name").lower()
+                == updated_configuration.get("name").lower()
+            ):
+                Found = True
+                for k, v in updated_configuration["CSV"].items():
+                    if v == "CSV":
+                        continue
+                    if data_source["CSV"].get(k) != v:
+                        data_source["CSV"][k] = v
+                        changed = True
+        if not Found:
+            configuration.get("DATA_SOURCES").append(updated_configuration)
+            changed = True
+    else:
+        for data_source in configuration.get("DATA_SOURCES"):
+            if data_source["name"].lower() == updated_configuration["name"].lower():
+                Found = True
+                for k, v in updated_configuration["DATABASE"].items():
+                    if data_source["DATABASE"][k] != v:
+                        data_source["DATABASE"][k] = v
+                        changed = True
+                break
+        if not Found:
+            configuration.get("DATA_SOURCES").append(updated_configuration)
+            changed = True
+    return changed
+
+
+def rename_datasource(data_source_name, new_data_source_name):
+    change = False
+    with open(app_config.INSTANCE_CONFIG) as f:
+        configuration = yaml.load(f)
+    for data_source in configuration.get("DATA_SOURCES"):
+        if data_source.get("name").lower() == data_source_name.lower():
+            data_source["name"] = new_data_source_name
+            change = True
+
+    if not change:
+        for data_source in configuration.get("DATA_SOURCES"):
+            if data_source["name"].lower() == data_source_name.lower():
+                data_source["name"] = new_data_source_name
+                return True
+    if change:
+        print(
+            "data source %s has been renamed to %s"
+            % (data_source_name, new_data_source_name)
+        )
+        with open(app_config.INSTANCE_CONFIG, "w") as f:
+            yaml.dump(configuration, f)
+
+
+def delete_data_source_configuration(data_source_name):
+    change = False
+    with open(app_config.INSTANCE_CONFIG) as f:
+        configuration = yaml.load(f)
+    for data_source in configuration.get("DATA_SOURCES"):
+        if data_source.get("name").lower() == data_source_name.lower():
+            configuration.get("DATA_SOURCES").remove(data_source)
+            change = True
+
+    if not change:
+        for data_source in configuration.get("DATA_SOURCES"):
+            if data_source["name"].lower() == data_source_name.lower():
+                configuration.get("DATA_SOURCES").remove(data_source)
+                return True
+    if change:
+        with open(app_config.INSTANCE_CONFIG, "w") as f:
+            yaml.dump(configuration, f)
+
+
+def get_configure_file():
+    home_folder = os.path.expanduser("~")

```

Comment:
Generally I don't think that the configuration file will be in the home directory

Reply:
In the development environment, the file is stored in this directory.

Reply:
I think we should review that part of the instruction later